File: lib/gemini/gemini.go
package gemini

import (
	"TextMatchCut/types"
	"TextMatchCut/util"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sync"

	"google.golang.org/genai"
)

var (
	geminiConfigInstance *genai.GenerateContentConfig
	geminiClientInstance *genai.Client
	configOnce           sync.Once
	clientOnce           sync.Once
	prevApiKey           string
)

func getGeminiConfig() *genai.GenerateContentConfig {
	configOnce.Do(func() {
		geminiConfigInstance = &genai.GenerateContentConfig{
			ResponseMIMEType: "application/json",
			ResponseSchema: &genai.Schema{
				Type: genai.TypeArray,
				Items: &genai.Schema{
					Type: genai.TypeObject,
					Properties: map[string]*genai.Schema{
						"text": {Type: genai.TypeString},
					},
				},
			},
		}
	})
	return geminiConfigInstance
}

// FIXME: could be implemented better
func getGeminiClient(ctx context.Context) (*genai.Client, error) {
	var err error
	clientOnce.Do(func() {
		geminiClientInstance, err = genai.NewClient(ctx, nil)
	})
	return geminiClientInstance, err
}

func GetSnippets(ctx context.Context, config types.Config) ([]types.TextSnippet, error) {
	if prevApiKey != config.ApiKey {
		os.Setenv("GEMINI_API_KEY", config.ApiKey)
		// reset client so it picks up new API key
		clientOnce = sync.Once{}
		prevApiKey = config.ApiKey
	}
	client, err := getGeminiClient(ctx)
	if err != nil {
		return nil, err
	}
	geminiConfig := getGeminiConfig()

	log.Printf("Prompt for AI: %s\n", config.Prompt)
	result, err := client.Models.GenerateContent(
		ctx,
		config.Model,
		genai.Text(config.Prompt),
		geminiConfig,
	)
	if err != nil {
		log.Printf("Error generating content: %v\n", err)
		return nil, err
	}

	var snippets []types.AITextSnippets
	if err := json.Unmarshal([]byte(result.Text()), &snippets); err != nil {
		return nil, err
	}

	aiSnippets := util.ParseSnippetsJSON(snippets, config.HighlightedText)

	fmt.Printf("Snippets look like this: %+v\n", aiSnippets)

	return aiSnippets, nil
}

File: frontend/src/lib/gemini.ts
//this file is needed because for some reason fetch requests do not work on wasm backend (go)
import { GoogleGenAI } from '@google/genai';
import { Config } from '@types';

import * as z from 'zod';

const snippetSchema = z.object({
  text: z.string(),
});

const snippetsArraySchema = z.array(snippetSchema);

export async function GetSnippetsGeminiWeb(
  config: Config,
  signal: AbortSignal
) {
  const ai = new GoogleGenAI({
    apiKey: config.ApiKey,
  });

  const res = await ai.models.generateContent({
    contents: config.Prompt,
    model: config.Model as string,
    config: {
      responseMimeType: 'application/json',
      responseJsonSchema: z.toJSONSchema(snippetsArraySchema),
      abortSignal: signal,
    },
  });

  if (signal.aborted) {
    return;
  }

  const final = snippetsArraySchema.parse(JSON.parse(res.text!));

  console.log('final snippets from gemini.ts', final);
  return final;
}
